### ISSUE TITLE
Nested Structs with custom Coercer

### ISSUE THREAD

**Issue by antoinesaliba** [maxveldink/sorbet-schema#88]
I didn't have access to make a branch and I got lazy so going to make this issue just so you have something to noodle on tomorrow if you're bored.

Sorbet-schema can be used to deserialize data into a struct with a custom coercer.
struct Example
  const :country, Country

It can also be used to deserialize data into a struct with a nested struct.
struct Nested
  const :name, String
  
struct Example
  const :nested, Nested
  
It doesn't work though when you want to deserialize data into a struct with a nested struct **that has a custom coercer.**
struct Nested
  const :country, Country
  
struct Example
  const :nested, Nested
  
With this last example, it'll still create the Example struct with the Nested struct correctly created but country will not be coerced, it'll just be the string representation I passed as json. This is because the nested struct will get converted using the Struct coercer and this will just not consider any custom coercers https://github.com/maxveldink/sorbet-schema/blob/main/lib/typed/coercion/struct_coercer.rb#L24.

Here's a first draft at how I got it working but I'm sure there's a better way to do it.
```
       sig { override.params(type: T::Types::Base, value: Value).returns(Result[Target, CoercionError]) }
      def coerce(type:, value:)
        return Failure.new(CoercionError.new("Field type must inherit from T::Struct for Struct coercion.")) unless used_for_type?(type)
        return Success.new(value) if type.recursively_valid?(value)

        return Failure.new(CoercionError.new("Value of type '#{value.class}' cannot be coerced to #{type} Struct.")) unless value.is_a?(Hash)

        values = {}
        type.raw_type.props.each do |name, prop|
          if prop[:type_object].try(:raw_type).respond_to?(:props)
            Typed::HashSerializer
              .new(schema: Typed::Schema.from_struct(prop[:type_object].raw_type))
              .deserialize(value[name])
              .and_then { |struct| Typed::Success.new(values[name] = struct) }
              .on_error { |error| return Failure.new(CoercionError.new("Nested hash for #{type} could not be coerced to #{name}")) }
          else
            Coercion
              .coerce(type: prop[:type_object], value: value[name])
              .and_then { |coerced_value| Typed::Success.new(values[name] = struct) }
              .on_error { |error| return Failure.new(CoercionError.new("Value for #{name} could not be coerced to #{prop[:type]}")) }
          end
        end

        Success.new(type.raw_type.new(values))
      rescue ArgumentError, RuntimeError
        Failure.new(CoercionError.new("Given hash could not be coerced to #{type}."))
      end
```

**Comment by maxveldink** [comment]
That looks close to what we need to do; I think this also fixes #71

**Comment by antoinesaliba** [comment]
This is resolved by https://github.com/maxveldink/sorbet-schema/pull/102
